File: backend/TradeX.Trading/Backtest/BacktestEngine.cs
using System.Text.Json;
using TradeX.Core.Interfaces;
using TradeX.Core.Models;
using TradeX.Indicators;
using TradeX.Trading.Rules;

namespace TradeX.Trading.Backtest;

public class BacktestEngine(IIndicatorRegistry indicators, StrategyEvaluator strategyEvaluator)
{
    private const int MaxKlines = 100_000;

    public (BacktestResult Result, List<BacktestTrade> Trades, List<BacktestKlineAnalysis> Analysis) Run(
        Strategy strategy,
        string pair,
        IReadOnlyList<Kline> klines,
        decimal initialCapital = 1000m,
        decimal? positionSize = null,
        Action<BacktestKlineAnalysis>? onAnalysis = null,
        string? timeframe = null,
        CancellationToken ct = default,
        decimal feeRate = 0m)
    {
        if (klines.Count < 50)
            return (CreateEmptyResult("数据不足，至少需要 50 根 K 线", initialCapital), [], []);

        if (klines.Count > MaxKlines)
            return (CreateEmptyResult($"数据量过大，超过 {MaxKlines} 根 K 线上限", initialCapital), [], []);

        List<BacktestTrade> trades = [];
        List<BacktestKlineAnalysis> analysis = [];
        var prices = klines.Select(c => c.Close).ToArray();
        var volumes = klines.Select(c => (long)c.Volume).ToArray();

        // 每次 Run 用唯一作用域键，隔离 MinInterval 冷却状态
        var scopeKey = $"backtest:{Guid.NewGuid():N}";

        // 历史指标快照列表
        List<Dictionary<string, decimal>> historicalSnapshots = [];

        var lots = new List<Lot>();
        var cash = initialCapital;
        List<decimal> equityCurve = [];

        decimal CloseLot(Lot lot, int exitIndex, Kline exitOhlc)
        {
            var exitPrice = exitOhlc.Close;
            var exitFee = lot.Quantity * exitPrice * feeRate;
            var pnl = (exitPrice - lot.EntryPrice) * lot.Quantity - lot.EntryFee - exitFee;
            var costBasis = lot.EntryPrice * lot.Quantity + lot.EntryFee;
            var pnlPercent = costBasis > 0 ? pnl / costBasis * 100 : 0;
            trades.Add(new BacktestTrade(
                lot.EntryIndex, exitIndex,
                lot.EntryTime, exitOhlc.Timestamp,
                lot.EntryPrice, exitPrice, lot.Quantity, pnl, pnlPercent));
            return lot.Quantity * exitPrice - exitFee;
        }

        // 创建回测用的 StrategyBinding
        var binding = new StrategyBinding
        {
            Id = Guid.NewGuid(),
            StrategyId = strategy.Id,
            TraderId = strategy.CreatedBy,
            Pairs = pair,
            Timeframe = timeframe ?? "1h",
            Status = Core.Enums.BindingStatus.Active,
            Name = strategy.Name,
        };

        for (var i = 50; i < prices.Length; i++)
        {
            ct.ThrowIfCancellationRequested();
            var ohlc = klines[i];
            var prevOhlc = klines[i - 1];
            var currentValues = indicators.ComputeAll(new KlineWindow(prices[..(i + 1)], volumes[..(i + 1)], ohlc.Open, ohlc.High, ohlc.Low, ohlc.Close));
            var previousValues = indicators.ComputeAll(new KlineWindow(prices[..i], volumes[..i], prevOhlc.Open, prevOhlc.High, prevOhlc.Low, prevOhlc.Close));

            var price = ohlc.Close;
            var quantityHeld = lots.Sum(l => l.Quantity);
            var avgEntry = quantityHeld > 0 ? lots.Sum(l => l.EntryPrice * l.Quantity) / quantityHeld : 0m;

            historicalSnapshots.Add(new Dictionary<string, decimal>(currentValues));

            // 通过 StrategyEvaluator 评估规则链
            if (strategy.Mode == Core.Enums.StrategyMode.RuleChain)
            {
                strategyEvaluator.EvaluateBindingChain(
                    binding, pair, price, Guid.Empty,
                    klines, ct);
            }

            // 末根强制平掉全部剩余持仓
            if (i == prices.Length - 1 && lots.Count > 0)
            {
                foreach (var lot in lots)
                    cash += CloseLot(lot, i, ohlc);
                lots.Clear();
            }

            var qtyAfter = lots.Sum(l => l.Quantity);
            var avgAfter = qtyAfter > 0 ? lots.Sum(l => l.EntryPrice * l.Quantity) / qtyAfter : 0m;

            analysis.Add(new BacktestKlineAnalysis(
                i, ohlc.Timestamp, ohlc.Open, ohlc.High, ohlc.Low, ohlc.Close, ohlc.Volume,
                currentValues, null, null, qtyAfter > 0, "none",
                avgAfter > 0 ? avgAfter : null,
                qtyAfter > 0 ? qtyAfter : null,
                qtyAfter > 0 ? avgAfter * qtyAfter : null,
                qtyAfter > 0 ? price * qtyAfter : null,
                qtyAfter > 0 ? (price - avgAfter) * qtyAfter : null,
                avgAfter > 0 ? (price - avgAfter) / avgAfter * 100m : null));

            equityCurve.Add(cash + qtyAfter * price);
            onAnalysis?.Invoke(analysis[^1]);
        }

        var finalEquity = equityCurve.Count > 0 ? equityCurve[^1] : initialCapital;
        var result = CalculateMetrics(trades, klines, equityCurve, initialCapital, finalEquity, timeframe);
        return (result, trades, analysis);
    }

    private sealed record Lot(decimal EntryPrice, decimal Quantity, decimal EntryFee, int EntryIndex, DateTime EntryTime);

    private static BacktestResult CalculateMetrics(
        IReadOnlyList<BacktestTrade> trades,
        IReadOnlyList<Kline> klines,
        IReadOnlyList<decimal> equityCurve,
        decimal initialCapital,
        decimal finalEquity,
        string? timeframe)
    {
        var jsonOptions = new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };

        if (trades.Count == 0)
            return new BacktestResult
            {
                InitialCapital = initialCapital,
                FinalValue = initialCapital,
                TotalReturnPercent = 0,
                AnnualizedReturnPercent = 0,
                MaxDrawdownPercent = 0,
                WinRate = 0,
                TotalTrades = 0,
                SharpeRatio = 0,
                ProfitLossRatio = 0,
                Details = "{\"message\":\"无交易产生\"}"
            };

        var wins = trades.Count(t => t.PnL > 0);
        var winRate = (decimal)wins / trades.Count * 100;

        var totalReturn = initialCapital > 0 ? (finalEquity - initialCapital) / initialCapital * 100 : 0;

        var totalDays = (klines[^1].Timestamp - klines[0].Timestamp).TotalDays;
        var annualizedReturn = 0m;
        if (totalDays > 0 && initialCapital > 0 && finalEquity > 0)
        {
            var growth = (double)(finalEquity / initialCapital);
            annualizedReturn = (decimal)Math.Max(-9999, Math.Min(9999, Math.Pow(growth, 365.0 / totalDays) - 1)) * 100;
        }

        var maxDrawdown = ComputeMaxDrawdown(equityCurve);
        var sharpe = ComputeSharpe(equityCurve, timeframe);

        var totalWin = trades.Where(t => t.PnL > 0).Sum(t => t.PnL);
        var totalLoss = trades.Where(t => t.PnL <= 0).Sum(t => Math.Abs(t.PnL));
        var profitLossRatio = totalLoss > 0 ? totalWin / totalLoss : totalWin > 0 ? totalWin : 0;

        var details = JsonSerializer.Serialize(trades.Select(t => new
        {
            t.EnteredAt, t.ExitedAt, t.EntryPrice, t.ExitPrice,
            t.Quantity, t.PnL, t.PnLPercent
        }), jsonOptions);

        return new BacktestResult
        {
            InitialCapital = initialCapital,
            FinalValue = Math.Round(finalEquity, 2),
            TotalReturnPercent = Math.Round(totalReturn, 2),
            AnnualizedReturnPercent = Math.Round(annualizedReturn, 2),
            MaxDrawdownPercent = Math.Round(maxDrawdown, 2),
            WinRate = Math.Round(winRate, 1),
            TotalTrades = trades.Count,
            SharpeRatio = Math.Round(sharpe, 2),
            ProfitLossRatio = Math.Round(profitLossRatio, 2),
            Details = details
        };
    }

    private static decimal ComputeMaxDrawdown(IReadOnlyList<decimal> equityCurve)
    {
        if (equityCurve.Count == 0) return 0m;
        var peak = equityCurve[0];
        var maxDrawdown = 0m;
        foreach (var equity in equityCurve)
        {
            if (equity > peak) peak = equity;
            if (peak > 0)
            {
                var dd = (peak - equity) / peak * 100;
                if (dd > maxDrawdown) maxDrawdown = dd;
            }
        }
        return maxDrawdown;
    }

    private static decimal ComputeSharpe(IReadOnlyList<decimal> equityCurve, string? timeframe)
    {
        if (equityCurve.Count < 3) return 0m;

        var returns = new List<double>(equityCurve.Count - 1);
        for (var i = 1; i < equityCurve.Count; i++)
        {
            var prev = (double)equityCurve[i - 1];
            if (prev <= 0) continue;
            returns.Add((double)equityCurve[i] / prev - 1);
        }
        if (returns.Count < 2) return 0m;

        var mean = returns.Average();
        var variance = returns.Average(r => Math.Pow(r - mean, 2));
        var stdDev = Math.Sqrt(Math.Max(0, variance));
        if (stdDev <= 0) return 0m;

        var periodsPerYear = PeriodsPerYear(timeframe);
        var sharpe = mean / stdDev * Math.Sqrt(periodsPerYear);
        return (decimal)Math.Max(-9999, Math.Min(9999, sharpe));
    }

    private static double PeriodsPerYear(string? timeframe) => timeframe switch
    {
        "1m" => 525_600,
        "5m" => 105_120,
        "15m" => 35_040,
        "30m" => 17_520,
        "1h" => 8_760,
        "4h" => 2_190,
        "1d" => 365,
        _ => 365
    };

    private static BacktestResult CreateEmptyResult(string reason, decimal initialCapital)
        => new()
        {
            InitialCapital = initialCapital,
            FinalValue = initialCapital,
            TotalReturnPercent = 0,
            AnnualizedReturnPercent = 0,
            MaxDrawdownPercent = 0,
            WinRate = 0,
            TotalTrades = 0,
            SharpeRatio = 0,
            ProfitLossRatio = 0,
            Details = $"{{\"message\":\"{reason}\"}}"
        };
}

File: backend-go/internal/domain/engine/backtest_engine.go
package engine

import (
	"context"
	"fmt"
	"math"
	"time"

	"github.com/shopspring/decimal"

	"tradex/internal/domain"
	"tradex/internal/domain/indicator"
)

const (
	MaxKlines       = 100000
	FirstValidIndex = 50
)

type EngineInput struct {
	Strategy       domain.Strategy
	Pair           string
	Klines         []domain.Candle
	InitialCapital decimal.Decimal
	PositionSize   *decimal.Decimal
	FeeRate        decimal.Decimal
	Timeframe      string
	OnAnalysis     func(domain.BacktestKlineAnalysis)
}

type EngineOutput struct {
	Result      domain.BacktestResult
	Trades      []domain.BacktestTrade
	Analysis    []domain.BacktestKlineAnalysis
	EquityCurve []decimal.Decimal
}

type BacktestEngine struct {
	registry *indicator.Registry
	decision *StrategyDecisionEngine
}

func NewBacktestEngine(registry *indicator.Registry) *BacktestEngine {
	return &BacktestEngine{
		registry: registry,
		decision: NewStrategyDecisionEngine(),
	}
}

func (e *BacktestEngine) Run(ctx context.Context, input EngineInput) (EngineOutput, error) {
	if err := e.validate(input); err != nil {
		return EngineOutput{}, err
	}

	trades := make([]domain.BacktestTrade, 0, 100)
	analysis := make([]domain.BacktestKlineAnalysis, 0, len(input.Klines))
	equityCurve := make([]decimal.Decimal, len(input.Klines))

	capital := input.InitialCapital
	equity := input.InitialCapital
	var inPosition bool
	var positionSize decimal.Decimal
	var avgEntryPrice decimal.Decimal
	var positionCount int
	var entryTimestamp time.Time

	e.registry.ComputeAll(input.Klines)

	for i := FirstValidIndex; i < len(input.Klines); i++ {
		select {
		case <-ctx.Done():
			return EngineOutput{}, ctx.Err()
		default:
		}

		candle := input.Klines[i]
		closePrice := candle.Close

		decision := e.decision.Decide(DecisionInput{
			Strategy:      input.Strategy,
			Klines:        input.Klines,
			Index:         i,
			Registry:      e.registry,
			InPosition:    inPosition,
			CurrentEquity: equity,
			CurrentPrice:  closePrice,
			PositionCount: positionCount,
			AvgEntryPrice: avgEntryPriceOrNil(inPosition, avgEntryPrice),
		})

		switch decision.Type {
		case DecisionEnter:
			posSize := input.PositionSize
			if decision.PositionSize != nil {
				posSize = decision.PositionSize
			}

			amount := posSize
			if amount == nil {
				amount = new(equity)
			}

			if amount.GreaterThan(equity) {
				amount = new(equity)
			}

			qty := amount.Div(closePrice)
			fee := amount.Mul(input.FeeRate)
			capital = capital.Sub(*amount).Sub(fee)

			if !inPosition {
				positionSize = qty
				avgEntryPrice = closePrice
				positionCount = 1
			} else {
				totalQty := positionSize.Add(qty)
				if totalQty.IsPositive() {
					avgEntryPrice = positionSize.Mul(avgEntryPrice).Add(qty.Mul(closePrice)).Div(totalQty)
				}
				positionSize = totalQty
				positionCount++
			}

			entryTimestamp = candle.Timestamp
			inPosition = true

		case DecisionExit:
			if inPosition {
				sellValue := positionSize.Mul(closePrice)
				fee := sellValue.Mul(input.FeeRate)
				pnl := positionSize.Mul(closePrice.Sub(avgEntryPrice))

				totalFee := fee
				capital = capital.Add(positionSize.Mul(closePrice)).Sub(totalFee)

				trade := domain.BacktestTrade{
					EntryIndex: 0,
					ExitIndex:  i,
					EnteredAt:  entryTimestamp,
					ExitedAt:   candle.Timestamp,
					EntryPrice: avgEntryPrice,
					ExitPrice:  closePrice,
					Quantity:   positionSize,
					PnL:        pnl,
					PnLPercent: pnl.Div(avgEntryPrice.Mul(positionSize)).Mul(decimal.NewFromInt(100)),
				}
				if len(trades) > 0 {
					prev := trades[len(trades)-1]
					trade.EntryIndex = prev.ExitIndex + 1
				}
				trades = append(trades, trade)

				equity = capital
				inPosition = false
				positionSize = decimal.Zero
				positionCount = 0
			}

		case DecisionHold:
			if inPosition {
				equity = capital.Add(positionSize.Mul(closePrice))
			} else {
				equity = capital
			}
		}

		equityCurve[i] = equity

		analysisEntry := domain.BacktestKlineAnalysis{
			KlineIndex: i,
			Timestamp:  candle.Timestamp,
			Open:       candle.Open,
			High:       candle.High,
			Low:        candle.Low,
			Close:      candle.Close,
			Volume:     candle.Volume,
			InPosition: inPosition,
			Action:     string(decision.Type),
		}
		if inPosition {
			analysisEntry.AvgEntryPrice = new(avgEntryPrice)
			analysisEntry.PositionQuantity = &positionSize
			analysisEntry.PositionValue = new(positionSize.Mul(closePrice))
			analysisEntry.PositionPnl = new(positionSize.Mul(closePrice.Sub(avgEntryPrice)))
		}

		indicatorValues := make(map[string]float64)
		for _, name := range e.registry.Indicators() {
			if val, ok := e.registry.GetValue(name, i); ok {
				indicatorValues[name] = val
			}
		}
		analysisEntry.IndicatorValues = indicatorValues

		analysis = append(analysis, analysisEntry)

		if input.OnAnalysis != nil {
			input.OnAnalysis(analysisEntry)
		}
	}

	result := e.computeResult(input, trades, equityCurve)
	return EngineOutput{
		Result:      result,
		Trades:      trades,
		Analysis:    analysis,
		EquityCurve: equityCurve,
	}, nil
}

func (e *BacktestEngine) validate(input EngineInput) error {
	if len(input.Klines) < FirstValidIndex+1 {
		return fmt.Errorf("not enough klines: got %d, need at least %d", len(input.Klines), FirstValidIndex+1)
	}
	if len(input.Klines) > MaxKlines {
		return fmt.Errorf("too many klines: got %d, max %d", len(input.Klines), MaxKlines)
	}
	if input.InitialCapital.IsNegative() || input.InitialCapital.IsZero() {
		return fmt.Errorf("initial capital must be positive")
	}
	return nil
}

func (e *BacktestEngine) computeResult(input EngineInput, trades []domain.BacktestTrade, equityCurve []decimal.Decimal) domain.BacktestResult {
	initialCapital := input.InitialCapital
	finalEquity := equityCurve[len(equityCurve)-1]

	totalReturn := decimal.Zero
	if initialCapital.IsPositive() {
		totalReturn = finalEquity.Sub(initialCapital).Div(initialCapital).Mul(decimal.NewFromInt(100))
	}

	duration := input.Klines[len(input.Klines)-1].Timestamp.Sub(input.Klines[FirstValidIndex].Timestamp)
	days := duration.Hours() / 24
	if days < 1 {
		days = 1
	}

	annualizedReturn := decimal.Zero
	if initialCapital.IsPositive() && !finalEquity.IsNegative() {
		ratio := finalEquity.Div(initialCapital)
		if ratio.IsPositive() {
			power := 365.0 / days
			ratioF64, _ := ratio.Float64()
			ratioF64 = math.Max(ratioF64, 0.0001)
			ratioF64 = math.Min(ratioF64, 1e10)
			annF64 := math.Pow(ratioF64, power) - 1
			if math.IsInf(annF64, 0) || math.IsNaN(annF64) {
				annF64 = 9999
			}
			annF64 = math.Max(annF64, -99.99)
			annF64 = math.Min(annF64, 9999)
			annualizedReturn = decimal.NewFromFloat(annF64).Mul(decimal.NewFromInt(100))
		}
	}

	maxDrawdown := e.computeMaxDrawdown(equityCurve)
	winRate := e.computeWinRate(trades)
	sharpe := e.computeSharpeRatio(equityCurve, days)
	profitLossRatio := e.computeProfitLossRatio(trades)

	return domain.BacktestResult{
		FinalValue:              finalEquity,
		TotalReturnPercent:      totalReturn,
		AnnualizedReturnPercent: annualizedReturn,
		MaxDrawdownPercent:      maxDrawdown,
		WinRate:                 winRate,
		SharpeRatio:             sharpe,
		ProfitLossRatio:         profitLossRatio,
		TotalTrades:             len(trades),
	}
}

func (e *BacktestEngine) computeMaxDrawdown(equityCurve []decimal.Decimal) decimal.Decimal {
	var peak decimal.Decimal
	var maxDD decimal.Decimal

	for _, eq := range equityCurve {
		if eq.IsZero() {
			continue
		}
		if eq.GreaterThan(peak) {
			peak = eq
		}
		if peak.IsPositive() {
			dd := eq.Sub(peak).Div(peak).Mul(decimal.NewFromInt(100))
			if dd.LessThan(maxDD) {
				maxDD = dd
			}
		}
	}

	return maxDD
}

func (e *BacktestEngine) computeWinRate(trades []domain.BacktestTrade) decimal.Decimal {
	if len(trades) == 0 {
		return decimal.Zero
	}
	wins := 0
	for _, t := range trades {
		if t.PnL.IsPositive() {
			wins++
		}
	}
	return decimal.NewFromInt(int64(wins)).Div(decimal.NewFromInt(int64(len(trades)))).Mul(decimal.NewFromInt(100))
}

func (e *BacktestEngine) computeSharpeRatio(equityCurve []decimal.Decimal, days float64) decimal.Decimal {
	returns := make([]float64, 0, len(equityCurve))
	prev := equityCurve[FirstValidIndex]
	for i := FirstValidIndex + 1; i < len(equityCurve); i++ {
		if prev.IsPositive() {
			ret, _ := equityCurve[i].Sub(prev).Div(prev).Float64()
			returns = append(returns, ret)
		}
		prev = equityCurve[i]
	}

	if len(returns) < 2 {
		return decimal.Zero
	}

	var sum, sumSq float64
	for _, r := range returns {
		sum += r
	}
	mean := sum / float64(len(returns))

	for _, r := range returns {
		d := r - mean
		sumSq += d * d
	}
	variance := sumSq / float64(len(returns)-1)
	std := math.Sqrt(variance)

	if std < 1e-15 {
		return decimal.Zero
	}

	periodsPerYear := 365.0 / days * float64(len(returns))
	sharpe := (mean / std) * math.Sqrt(periodsPerYear)

	return decimal.NewFromFloat(sharpe)
}

func (e *BacktestEngine) computeProfitLossRatio(trades []domain.BacktestTrade) decimal.Decimal {
	totalProfit := decimal.Zero
	totalLoss := decimal.Zero

	for _, t := range trades {
		if t.PnL.IsPositive() {
			totalProfit = totalProfit.Add(t.PnL)
		} else {
			totalLoss = totalLoss.Add(t.PnL.Abs())
		}
	}

	if totalLoss.IsZero() {
		if totalProfit.IsZero() {
			return decimal.Zero
		}
		return decimal.NewFromInt(999)
	}

	return totalProfit.Div(totalLoss)
}

func avgEntryPriceOrNil(inPosition bool, price decimal.Decimal) *decimal.Decimal {
	if inPosition {
		return &price
	}
	return nil
}
